---
title: "PNA_RetSeq"
author: "Jakob Jung"
date: "2/5/2021"
output: html_document
---

Here I will analyze the ribosome profiling data of our oligo-pool experiment, in which we challenged the whole transcriptome of *E. coli* with several transcript-specific PNA. 

# Package import

Here, I import packages and the generated count tables:
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Packages:
```{r}
library(ggplot2)
library(dplyr)
library(hrbrthemes)
library(viridis)
library(tidyr)
library(edgeR)
library(tidybulk)
library(ggpubr)
```




# Maping statistics

## percentages of mapped read types
Here I show the distribution of different reads (oligos, rRNA, tRNA) to get an idea on how the rRNA depletion worked out.
First I import total counts per rRNA, mRNA and tRNA and look at the total counts of our ologos:
```{r}
oligos_totcounts <- unlist(read.delim("../../pna_retseq_11_2020/data/rna_align/count_table_genenames.txt.summary", 
                                      row.names = 1)[1,])
rrna_totcounts <- unlist(read.delim("../../pna_retseq_11_2020/data/rna_align/count_table_genenames.txt.summary", 
                                    row.names = 1)[1,])
oligos_totcounts
```

Now I create a dataframe and plot a stacked barplot of the data: 
```{r}
df_rnatypes <- data.frame(rna_type = c(rep("rRNA", length(rrna_totcounts)), rep("mRNA", length(rrna_totcounts))),
                          counts = c(rrna_totcounts, rrna_totcounts), samples = rep(names(rrna_totcounts), 2))

g <- ggplot(df_rnatypes) + aes(x=samples, y=counts, fill = rna_type) + 
  geom_bar(stat = "identity") + theme_ipsum_pub() + scale_fill_viridis(discrete = T)
g
```



## read length distribution
```{r}
path = "~/Documents/riboseq_experiments/pna_retseq_11_2020/analysis/readlengths/readlengths_after/"
file.names <- dir(path, pattern =".txt")
file.names.tot <- paste(path, file.names, sep = "")

sampless <- gsub(".*-\\d\\d-(.*)_S\\d.*", "\\1", file.names)
samples <- gsub("(.*)_S\\d.*", "\\1", sampless)
samples <- gsub("-", "_", samples)

for (i in 1:length(file.names.tot)) {
  f <- read.delim(file.names.tot[i], header = F, sep = "\t")
  
  names(f) <- c("count","readlength")
  g <- ggplot(f) + aes(x=readlength, y=count) + 
    geom_bar(stat = "identity",width = 0.8) + theme_ipsum_pub() + 
    scale_y_continuous(name="read counts", labels = scales::comma) +
    ggtitle(samples[i]) + scale_x_continuous(breaks = seq(10, 75, by = 10))
  print(g)
}
```


# Differential analysis of oligo pool:

## Data import
I start with importing count data (genewisecounts):
```{r}
# import:
GenewiseCounts <- read.delim(
    "../../../pna_rnaseq_experiments/pna_rnaseq_11_20/data/rna_align/counttable.txt",
    sep = "\t", header = T, comment.char = "#")
dim(GenewiseCounts)
head(GenewiseCounts)
```

I move on with adding sample names etc.:
```{r}
gwc <- GenewiseCounts[,c(1,8:length(GenewiseCounts[1,])-1)]
pnapat <- ".*\\.(later|beginning)\\.(ctrl|[0-9]*nM)?.*\\.bam"
colnames (gwc) <- gsub(pnapat,"\\1\\2", colnames(gwc))
colnames (gwc) <- gsub(".*(Undetermined).*","\\1", colnames(gwc))
#gwc <- gwc[,-1]
test <- factor(colnames(gwc))
# just for testing purposes!!
gwc <- gwc[, c(1,2,12,22,3,13,23,4,14,24)] 
colnames(gwc) <- c("transcript", "sample1_R1", "sample1_R2", "sample1_R3", "sample2_R1", "sample2_R2", 
                   "sample2_R3", "sample3_R1", "sample3_R2", "sample3_R3")
head(gwc)
```

combine dataframes to tidybulk tibble:
```{r}
colnames(gwc)[1] <- "transcript"

# create raw counts and add condition variable:
raw_counts <- pivot_longer(gwc, cols = c(2:length(gwc[1,])), names_to = "sample", values_to = "counts") 
raw_counts$cond <- gsub("(sample\\d+).*", "\\1" ,  raw_counts$sample )

# create tidybulk tibble (only add samples, transcripts and counts!):
tt <- raw_counts %>% tidybulk(sample, transcript, counts)

# aggregate samples (remove duplicate rows, add them):
tt.aggr <-  tt %>% aggregate_duplicates( aggregation_function = sum )
tt.aggr
```



## Normalization

Now I normalize the counts using TMM (automaticaly ignores lowly abundant transcripts):
```{r}
tt.norm <- tt.aggr %>%  scale_abundance(method = "TMM", factor_of_interest = cond) 
```
Now I create  a plot of count distributions:
```{r}
tt.norm %>%
  pivot_longer(         
    c(counts, counts_scaled),         
    values_to="count",          
    names_to="Normalisation"
    ) %>%     
  ggplot(aes(count + 1, group=cond, color=cond)) +     
  facet_grid(~Normalisation) +     
  geom_density() +     
  scale_x_log10() + theme_ipsum_pub()

```

## Dimensionality reduction
Do MDS:
```{r}
tt.norm.mds <- tt.norm %>% reduce_dimensions(.abundance = counts_scaled, method="MDS", .dims = 3)

tt.norm.mds %>%
  pivot_sample() %>%
  GGally::ggpairs(columns = 6:7, ggplot2::aes(colour=cond)) + theme_ipsum_pub()

tt.norm.mds %>%
  pivot_sample() %>%
  ggplot(aes(x=Dim1, y=Dim2, size = 1, color=cond)) + geom_point() + theme_ipsum_pub()
```

Do PCA:
```{r}
tt.norm.pca <- tt.norm %>% reduce_dimensions(.abundance = counts_scaled, method="PCA", .dims = 3)

tt.norm.pca %>%
  pivot_sample() %>%
  GGally::ggpairs(columns = 6:7, ggplot2::aes(colour=cond)) + theme_ipsum_pub()

tt.norm.pca %>%
  pivot_sample() %>%
  ggplot(aes(x=PC1, y=PC2, size = 1, color=cond)) + geom_point() + theme_pubr()
```


Now we test for differential expression:
```{r}
  tt %>% test_differential_abundance( ~ cond, action = "only")
```















